x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/VaragsTestCases.java

```diff
@@ -0,0 +1,160 @@
+ */
+public class VaragsTestCases {
+    public static List<AbstractFunctionTestCase.TestCaseSupplier> anyNullIsNull(
+        Function<String, String> expectedEvaluatorPrefix,
+        Function<Stream<String>, String> expectedStr,
+        Function<LongStream, OptionalLong> expectedLong,
+        Function<IntStream, OptionalInt> expectedInt
+    ) {
+        return new VaragsTestCases(expectedEvaluatorPrefix, strings -> {
+            if (Arrays.stream(strings).anyMatch(s -> s == null)) {
+                return nullValue();
+            }
+            return equalTo(new BytesRef(expectedStr.apply(Arrays.stream(strings))));
+        }, longs -> {
+            if (Arrays.stream(longs).anyMatch(l -> l == null)) {
+                return nullValue();
+            }
+            return equalTo(expectedLong.apply(Arrays.stream(longs).mapToLong(Long::longValue)).getAsLong());
+        }, ints -> {
+            if (Arrays.stream(ints).anyMatch(i -> i == null)) {
+                return nullValue();
+            }
+            return equalTo(expectedInt.apply(Arrays.stream(ints).mapToInt(Integer::intValue)).getAsInt());
+        }).suppliers();
+    }
+
+    private static final int MAX_WIDTH = 10;
+
+    private final List<AbstractFunctionTestCase.TestCaseSupplier> suppliers = new ArrayList<>();
+
+    private final Function<String, String> expectedEvaluatorPrefix;
+    private final Function<String[], Matcher<Object>> expectedStr;
+    private final Function<Long[], Matcher<Object>> expectedLong;
+    private final Function<Integer[], Matcher<Object>> expectedInt;
+
+    public VaragsTestCases(
+        Function<String, String> expectedEvaluatorPrefix,
+        Function<String[], Matcher<Object>> expectedStr,
+        Function<Long[], Matcher<Object>> expectedLong,
+        Function<Integer[], Matcher<Object>> expectedInt
+    ) {
+        this.expectedEvaluatorPrefix = expectedEvaluatorPrefix;
+        this.expectedStr = expectedStr;
+        this.expectedLong = expectedLong;
+        this.expectedInt = expectedInt;
+    }
+
+    public List<AbstractFunctionTestCase.TestCaseSupplier> suppliers() {
+        // TODO more types
+        strings();
+        longs();
+        ints();
+        return suppliers;
+    }
+
+    private void strings() {
+        IntStream.range(1, MAX_WIDTH)
+            .mapToObj(i -> new AbstractFunctionTestCase.TestCaseSupplier(testCaseName(i, "kwd"), () -> strings(DataTypes.KEYWORD, i)))
+            .forEach(suppliers::add);
+        IntStream.range(1, MAX_WIDTH)
+            .mapToObj(i -> new AbstractFunctionTestCase.TestCaseSupplier(testCaseName(i, "txt"), () -> strings(DataTypes.TEXT, i)))
+            .forEach(suppliers::add);
+    }
+
+    private AbstractFunctionTestCase.TestCase strings(DataType dataType, int i) {
+        String[] data = IntStream.range(0, i).mapToObj(v -> ESTestCase.randomAlphaOfLength(5)).toArray(String[]::new);
+        int[] n = new int[1];
```

Comment:
question: why do we use a single-element array here rather than a simple int?

Reply:
In java when you close over something which I do below it has to be "effectively final", meaning that the bytes of the thing you are directly pointing to don't change. If I did `int n` then I couldn't mutate `n` in the closure below. But and array is fine, because when  you close over an array you close over the reference, not the number itself. So you can mutate the contents. It's a sneaky java-ism that doesn't belong in high performance code or any other language with other constraints.

Reply:
Thanks, I wasn't aware!